micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java

```diff
@@ -607,14 +619,15 @@ public <T extends Collection<?>> T gaugeCollectionSize(String name, Iterable<Tag
     }
 
     private <M extends Meter> M registerMeterIfNecessary(Class<M> meterClass, Meter.Id id,
-            Function<Meter.Id, M> builder, Function<Meter.Id, M> noopBuilder) {
-        return registerMeterIfNecessary(meterClass, id, null, (id2, conf) -> builder.apply(id2), noopBuilder);
+            BiFunction<MeterRegistry, Meter.Id, M> meterSupplier, Function<Meter.Id, M> noopBuilder) {
+        return registerMeterIfNecessary(meterClass, id, null, null,
+                (registry, mappedId, conf, pd) -> meterSupplier.apply(registry, mappedId), noopBuilder);
```

Comment:
This adapting of the `BiFunction` to `NewMeterSupplier` is a capturing lambda that allocates (it did before these changes too), which is why we avoid calling this variant of `registerMeterIfNecessary` from the Counter call stack. It's convenient though, so we still use it for Function meters that should only ever be registered once anyway.